proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,381 @@
+  associatedtype Element where AsyncIterator.Element == Element
+  associatedtype Failure where AsyncIterator.Failure == Failure
+  func makeAsyncIterator() -> AsyncIterator
+}
+```
+
+This allows the use of `AsyncSequence` with both opaque types (`some
+AsyncSequence<String, any Error>`) and existential types (`any
+AsyncSequence<Image, NetworkError>`). 
+
+### AsyncThrowingStream
+
+The non-throwing variant `AsyncStream` cannot adopt typed throws
+since it does not have a generic `Failure` parameter; however,
+the throwing variant `AsyncThrowingStream` already contains a `Failure`
+generic parameter and can be extended to adopt typed throws.
+The current initializers of `AsyncThrowingStream` are constraining
+the `Failure` type to `Error`.
+
+```swift
+public init(
+  _ elementType: Element.Type = Element.self,
+  bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded,
+  _ build: (Continuation) -> Void
+) where Failure == Error
+
+public init(
+  unfolding produce: @escaping () async throws -> Element?
+) where Failure == Error 
+```
+
+We propose to add two new initializers that drop the `Failure` constraint:
+
+```swift
+public init(
+  of elementType: Element.Type = Element.self,
+  throwing failureType: Failure.Type = Failure.self,
+  bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded,
+  _ build: (Continuation) -> Void
+)
+
+public init(
+  unfolding produce: @escaping () async throws(Failure) -> Element?
+)
+```
+
+Furthermore, we propose to add a new `makeStream` method that
+similarly drop the `Failure` constraint:
+
+```swift
+@backDeployed(before: SwiftStdlib 5.9)
+public static func makeStream(
+    of elementType: Element.Type = Element.self,
+    throwing failureType: Failure.Type = Failure.self,
+    bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded
+) -> (stream: AsyncThrowingStream<Element, Failure>, continuation: AsyncThrowingStream<Element, Failure>.Continuation) {
+  var continuation: AsyncThrowingStream<Element, Failure>.Continuation!
+  let stream = AsyncThrowingStream<Element, Failure>(bufferingPolicy: limit) { continuation = $0 }
+  return (stream: stream, continuation: continuation!)
+}
+```
+
+### Throwing transformation sequences
+
+The `Concurrency` module contains a few algorithms for `AsyncSequence`. Those
+algorithms are backed by concrete types and have both a throwing and
+non-throwing variant. The throwing types do not contain a generic `Failure`
+parameter so we cannot adopt typed throws on them. However, we propose to add
+new underlying types with a generic `Failure` parameter and use opaque result
+types. Importantly though those new methods cannot be back-deployed since they
+introduce new types. Those are the new typed throw APIs that we propose to add:
+
+```swift
+extension AsyncSequence {
+  public func map<Transformed, Failure: Error>(
+    _ transform: @Sendable @escaping (Element) async throws -> Transformed
```

Comment:
the `throws` here is `any Error` so therefore the return type cannot be `Failure` - map has a potential of two possible failures: the downstream `Failure` plus the failure of the transform.